### PR TITLE
Client Encryption Bulk: Adds the ability to pass container rid to SDK header for Bulk operations for container recreate scenarios

### PR DESCRIPTION
**Pull request by kr-santosh**
## Description

This PR adds a fix to pass Encryption specific headers  passed as part of RequestOptions AddRequestHeaders delegate during bulk operation. This fix is required to handle the issue which was fixed by this PR: https://github.com/Azure/azure-cosmos-dotnet-v3/pull/2403 .

## Type of change
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs [hunk 1]
```diff
@@ -52,9 +64,18 @@ internal sealed class PartitionKeyRangeServerBatchRequest : ServerBatchRequest
             int maxOperationCount,
             bool ensureContinuousOperationIndexes,
             CosmosSerializerCore serializerCore,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken,
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:67]: CancellationToken is usually the last prameter.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs [hunk 2]
```diff
@@ -121,6 +122,114 @@ public async Task CreateItemAsync_WithBulk()
...
+            }
+
+            try
+            {
+                await Task.WhenAll(tasks);
+                Assert.Fail("Bulk execution should have failed. ");
+            }
+            catch(CosmosException ex)
+            {
+                if(ex.StatusCode == HttpStatusCode.Created || ex.SubStatusCode != 1024)
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs:166]: Why Created status code?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs [hunk 3]
```diff
@@ -121,6 +122,114 @@ public async Task CreateItemAsync_WithBulk()
             }
         }
 
+        [TestMethod]
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs:125]: Can we also add test for PartitionKeyRangeServerBatchRequest?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs [hunk 4]
```diff
@@ -166,24 +174,22 @@ public virtual bool TryAdd(ItemBatchOperation operation)
                         timeTakenInMilliseconds: stopwatch.ElapsedMilliseconds,
                         numberOfThrottles: numThrottle);
 
-                    using (PartitionKeyRangeBatchResponse batchResponse = new PartitionKeyRangeBatchResponse(serverRequest.Operations.Count, result.ServerResponse, this.serializerCore))
+                    PartitionKeyRangeBatchResponse batchResponse = new PartitionKeyRangeBatchResponse(serverRequest.Operations.Count, result.ServerResponse, this.serializerCore);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs:177]: Why this change of removing the scope?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 5]
```diff
@@ -88,10 +88,32 @@ internal BatchAsyncContainerExecutor()
...
+                itemRequestOptions.AddRequestHeaders?.Invoke(encryptionHeaders);
+
+                // make sure we set the Intended Collection Rid header when we have encrypted payload.
+                // This primarily would allow CosmosDB Encryption package to detect change in container referenced by a Client
+                // and prevent creating data with wrong Encryption Policy.
+                if (encryptionHeaders.TryGetValue(HttpConstants.HttpHeaders.IsClientEncrypted, out string encrypted))
+                {
+                    context.isClientEncrypted = bool.Parse(encrypted);
+
+                    if (encryptionHeaders.TryGetValue("x-ms-cosmos-intended-collection-rid", out string ridValue))
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:110]: Need to do this only if encrypted value is true.

Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:110]: Use WFConstants.BackendHeaders.IntendedCollectionRid

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs [hunk 6]
```diff
@@ -121,6 +122,114 @@ public async Task CreateItemAsync_WithBulk()
...
+
+            string rid = await containerInternal.GetCachedRIDAsync(forceRefresh: false, NoOpTrace.Singleton, cancellationToken: default);
+
+            // case 1. use wrong rid by using a stale rid.
+            ItemRequestOptions itemRequestOptions = new ItemRequestOptions()
+            {
+                AddRequestHeaders = (headers) =>
+                {
+                    headers["x-ms-cosmos-is-client-encrypted"] = bool.TrueString;
+                    headers["x-ms-cosmos-intended-collection-rid"] = rid;
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs:142]: Use Constants instead of strings

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 7]
```diff
@@ -176,9 +198,16 @@ public void Dispose()
...
-        private static void AddHeadersToRequestMessage(RequestMessage requestMessage, string partitionKeyRangeId)
+        private static void AddHeadersToRequestMessage(RequestMessage requestMessage, PartitionKeyRangeServerBatchRequest partitionKeyRangeServerBatchRequest)
         {
-            requestMessage.Headers.PartitionKeyRangeId = partitionKeyRangeId;
+            requestMessage.Headers.PartitionKeyRangeId = partitionKeyRangeServerBatchRequest.PartitionKeyRangeId;
+
+            if (partitionKeyRangeServerBatchRequest.isClientEncrypted)
+            {
+                requestMessage.Headers.Add(HttpConstants.HttpHeaders.IsClientEncrypted, partitionKeyRangeServerBatchRequest.isClientEncrypted.ToString());
+                requestMessage.Headers.Add("x-ms-cosmos-intended-collection-rid", partitionKeyRangeServerBatchRequest.intendedCollectionRidValue);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:208]: Use WFConstants.BackendHeaders.IntendedCollectionRid

File: Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperationContext.cs [hunk 8]
```diff
@@ -18,6 +18,10 @@ internal class ItemBatchOperationContext : IDisposable
     {
         public string PartitionKeyRangeId { get; private set; }
 
+        public bool isClientEncrypted { get; set; }
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperationContext.cs:21]: set default to false?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs [hunk 9]
```diff
@@ -224,7 +230,9 @@ public virtual bool TryAdd(ItemBatchOperation operation)
                   this.maxBatchOperationCount,
                   ensureContinuousOperationIndexes: false,
                   serializerCore: this.serializerCore,
-                  cancellationToken: cancellationToken).ConfigureAwait(false);
+                  cancellationToken: cancellationToken,
+                  isClientEncrypted: this.isClientEncrypted,
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs:234]: Just to confirm - we are not using batchOperations[0].Context.IsClientEncrypted directly here because it wont be set for Delete operation.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 5 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:110]: ah i think it was updated...it was not available earlier..

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs (hunk 6 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs:142]: tracking it.. need to followup and add "x-ms-cosmos-is-client-encrypted" to the header constants..

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 5 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:110]: ah i think it was updated...it was not available earlier..

File: Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs (hunk 1 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:67]: dont want to make cancellation token optional..

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs (hunk 4 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs:177]: was losing the response..

File: Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperationContext.cs (hunk 8 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperationContext.cs:21]: bool is false by default

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs (hunk 9 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs:234]: yes. So there can be operations like Read/Delete which may not carry the header..

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs (hunk 2 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs:166]: Either ways...

File: Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs (hunk 1 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:67]: Do we need to make the new parameters optional?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs (hunk 3 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Batch/CosmosItemBulkTests.cs:125]: The server request are built out of the batch request...are part of PartitionKeyRangeServerBatchRequest. It flows through that path.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs (hunk 4 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs:177]: Where are you using the batchResponse? Also, we should dispose it after use.

File: Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperationContext.cs (hunk 8 above)
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperationContext.cs:21]: >ool isC [](http://example.com/codeflow?start=16&length=7)

Capital i for both

File: Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs [hunk 10]
```diff
@@ -14,36 +14,48 @@ internal sealed class PartitionKeyRangeServerBatchRequest : ServerBatchRequest
...
+            this.isClientEncrypted = isClientEncrypted;
+            this.intendedCollectionRidValue = intendedCollectionRidValue;
         }
 
         /// <summary>
         ///  Gets the PartitionKeyRangeId that applies to all operations in this request.
         /// </summary>
         public string PartitionKeyRangeId { get; }
 
+        public bool isClientEncrypted { get; }
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:41]: >ol isCl [](http://example.com/codeflow?start=17&length=7)

capital i in both places

**Review by abhijitpai — DISMISSED** [review]
:shipit:

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs (hunk 4 above)
Comment (abhijitpai) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs:177]: We discussed this change is not needed.

File: Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs (hunk 1 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:67]: Since this might cause confusion for someone who has no idea about encryption package, using default/optional makes sense.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 11]
```diff
@@ -88,10 +88,32 @@ internal BatchAsyncContainerExecutor()
...
+                itemRequestOptions.AddRequestHeaders?.Invoke(encryptionHeaders);
+
+                // make sure we set the Intended Collection Rid header when we have encrypted payload.
+                // This primarily would allow CosmosDB Encryption package to detect change in container referenced by a Client
+                // and prevent creating data with wrong Encryption Policy.
+                if (encryptionHeaders.TryGetValue(HttpConstants.HttpHeaders.IsClientEncrypted, out string encrypted))
+                {
+                    context.IsClientEncrypted = bool.Parse(encrypted);
+
+                    if (context.IsClientEncrypted && encryptionHeaders.TryGetValue(WFConstants.BackendHeaders.IntendedCollectionRid, out string ridValue))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:110]: Why do we restrict this to only populate if the client is encrypted? Shouldn't it always just populate the intended collection rid value?

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:110]: The default behaviour we dont really pass this value.

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:110]: and for the flag IsClientEncrypted  the BE checks for this flag only for write operations...

File: Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs [hunk 12]
```diff
@@ -52,9 +64,18 @@ internal sealed class PartitionKeyRangeServerBatchRequest : ServerBatchRequest
             int maxOperationCount,
             bool ensureContinuousOperationIndexes,
             CosmosSerializerCore serializerCore,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken,
+            bool isClientEncrypted = false,
+            string intendedCollectionRidValue = null)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:69]: nit: not a blocker but it's usually better to avoid optional parameters. It avoids bugs caused by devs forgetting to pass the parameter through.

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:69]: the reason I made it optional is, for dev who are not aware of encryption package might find it confusing hence, having a default would make it much easier.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Batch/PartitionKeyRangeServerBatchRequest.cs:69]: Any dev working on this code should know what is and the dependency is has on the SDK. If they don't they can easily cause a regression. To me it's better for them to ask and understand what this is then to just assume it doesn't apply to what ever scenario they are working on.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 13]
```diff
@@ -88,10 +88,32 @@ internal BatchAsyncContainerExecutor()
...
+                // get the header value if any, passed by the encryption package.
+                Headers encryptionHeaders = new Headers();
+                itemRequestOptions.AddRequestHeaders?.Invoke(encryptionHeaders);
+
+                // make sure we set the Intended Collection Rid header when we have encrypted payload.
+                // This primarily would allow CosmosDB Encryption package to detect change in container referenced by a Client
+                // and prevent creating data with wrong Encryption Policy.
+                if (encryptionHeaders.TryGetValue(HttpConstants.HttpHeaders.IsClientEncrypted, out string encrypted))
+                {
+                    context.IsClientEncrypted = bool.Parse(encrypted);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:108]: Can some operations be encrypted and others not? If not, would it make sense to check each request?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 14]
```diff
@@ -176,9 +198,16 @@ public void Dispose()
...
-        private static void AddHeadersToRequestMessage(RequestMessage requestMessage, string partitionKeyRangeId)
+        private static void AddHeadersToRequestMessage(RequestMessage requestMessage, PartitionKeyRangeServerBatchRequest partitionKeyRangeServerBatchRequest)
         {
-            requestMessage.Headers.PartitionKeyRangeId = partitionKeyRangeId;
+            requestMessage.Headers.PartitionKeyRangeId = partitionKeyRangeServerBatchRequest.PartitionKeyRangeId;
+
+            if (partitionKeyRangeServerBatchRequest.IsClientEncrypted)
+            {
+                requestMessage.Headers.Add(HttpConstants.HttpHeaders.IsClientEncrypted, partitionKeyRangeServerBatchRequest.IsClientEncrypted.ToString());
+                requestMessage.Headers.Add(WFConstants.BackendHeaders.IntendedCollectionRid, partitionKeyRangeServerBatchRequest.IntendedCollectionRidValue);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:208]: Do we document anywhere what does the user need to do if he gets the 1024 sub status code? Or is this caught by the Encryption library?

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:208]: this is handled by Encryption library, the idea around this is 1) IsClientEncrypted - this sets the flag for BE validation of encrypted data if an encryption policy is configured. 2) IntendedCollectionRid/1024 error allows us to detect if the container was recreated with a different encryption policy , this allows the client to refresh its cached policy.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 13 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:108]: Yes, even w.r.t Encryption package we don't send out the header for operations like read or delete which don't modify data. But I don't see a way around it, however we can may be have an additional check to prevent the object creation and check for if (itemRequestOptions != null && itemRequestOptions.AddRequestHeaders != null)

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs [hunk 15]
```diff
@@ -88,10 +88,32 @@ internal BatchAsyncContainerExecutor()
                 trace, 
                 cancellationToken).ConfigureAwait(false);
             BatchAsyncStreamer streamer = this.GetOrAddStreamerForPartitionKeyRange(resolvedPartitionKeyRangeId);
+
             ItemBatchOperationContext context = new ItemBatchOperationContext(
                 resolvedPartitionKeyRangeId,
                 trace,
                 BatchAsyncContainerExecutor.GetRetryPolicy(this.cosmosContainer, operation.OperationType, this.retryOptions));
+
+            if (itemRequestOptions != null)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:97]: nit: Avoid the header logic if the function is null 
itemRequestOptions != null && temRequestOptions.AddRequestHeaders != null

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:97]: was about to add this.

**Review by j82w — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs (hunk 13 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:108]: Operations (that modify data - create, upsert, replace, patch ) the encryption package always adds the headers so any of these operation were to be part of the bulk operation we will add the header to the enter server request. Because the BE will check for the flag(only if the container has an encryption policy set) for these operations and fail the entire request. The reason why we need to bring this change is because in case of Bulk execution a new request option is built and sent to the BE as part of the server request. This PR basically makes sure that if any of the point operation which is part of bulk has a request option with the header set it  adds it  to bulk request option. So if its an operation which does not modify(and no header) being part of a batch operation which have operations with header does not affect it.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:108]: The issues is if the user sends 100 items with bulk enabled. 1 item has encryption and the others do not. The 99 items will fail because of the 1 encryption item. Will the 99 items automatically be retried in the encryption SDK? I think the additional latency should be acceptable with it being a corner cases during container recreates.

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs:108]: There is a PR(https://github.com/Azure/azure-cosmos-dotnet-v3/pull/2648) in progress which will remove this retry, also the current retry logic depends on specific sub status code which we will not hit in case.

**Review by ealsur — APPROVED** [review]

**Review by j82w — APPROVED** [review]
